### PR TITLE
test(slider): add unit tests to Slider

### PR DESCRIPTION
**Pull request by blackswanny**
add unit tests to Slider

### CODE REVIEW

File: src/slider/__tests__/slider.test.js [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+      step: 0,
+    };
+  });
+
+  afterEach(function() {
+    jest.restoreAllMocks();
+    document.addEventListener.mockClear();
+    wrapper && wrapper.unmount();
+    //$FlowFixMe
+    document.addEventListener = originalDocumentAddListener;
```
Comment (nadiia) [inline src/slider/__tests__/slider.test.js]: We don't need to restore it after every single test but can do it in the `afterAll` instead

File: src/slider/__tests__/slider.test.js [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+      range: Array<number>,
+      value: Array<number>,
+      step: ?number,
+      onChange: () => void,
+    },
+    mockFn;
+
+  beforeEach(function() {
+    //$FlowFixMe
+    document.addEventListener = jest.fn();
```
Comment (nadiia) [inline src/slider/__tests__/slider.test.js]: We clear mock after each test therefore we can do this assignment only once in the `beforeAll`

File: src/slider/__tests__/slider.test.js [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+      step = 10;
+      axisSizeInPixels = 300;
+      jest
+        .spyOn(window, 'getComputedStyle')
+        .mockImplementation(() => ({width: axisSizeInPixels}));
+      allProps.step = step;
+    });
+
+    it('should start tracking thumb move when user clicks on it', function() {
+      emulateThumbDown();
```
Comment (nadiia) [inline src/slider/__tests__/slider.test.js]: here we should also check that it subscribed for the `mousemove` and `mouseup` events

File: src/slider/__tests__/slider.test.js [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+      it('should reset current move progress when thumb moved to X pixels', function() {
+        emulateThumbDown();
+        event.movementX = 50;
+        fireMouseMove(event);
+        expect(wrapper.instance().state).toMatchObject({
+          currentMove: 0,
+        });
+      });
+      it('should note move thumb to X pixels if mouse moved less then step', function() {
+        allProps.step = 50;
```
Comment (nadiia) [inline src/slider/__tests__/slider.test.js]: It seems strange that we modify the shared `allProps` object here. I can see that we override it before each test, but the code looks messy. How about making `emulateThumbDown` accept props override, for example?

File: src/slider/__tests__/slider.test.js [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        emulateThumbDown();
+        event.movementX = 20;
+        fireMouseMove(event);
+        expect(wrapper.instance().state).toMatchObject({
+          currentMove: 20,
```
Comment (nadiia) [inline src/slider/__tests__/slider.test.js]: Could you explain why currentMove is expected to be set in this case but not in the test case above with almost the same setup?

File: src/slider/__tests__/stateful-slider.test.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+// @flow
+import React from 'react';
+import {mount} from 'enzyme';
+
+describe('Stateful slider', function() {
+  let allProps: any, wrapper;
+
+  beforeEach(function() {
+    allProps = {};
+    jest.mock('../slider', () => jest.fn(() => <div>test</div>));
```
Comment (nadiia) [inline src/slider/__tests__/stateful-slider.test.js]: why do we need to mock the Slider here?

File: src/slider/__tests__/slider.test.js (hunk 1 above)
Comment (blackswanny) [reply src/slider/__tests__/slider.test.js]: restore it in `afterAll` is not safe to preserve isolation of unit-test and may cause conflicts in environments. We already have some code in other places, which does similar

Comment (blackswanny) [reply src/slider/__tests__/slider.test.js]: ok, moved to afterAll

File: src/slider/__tests__/slider.test.js (hunk 2 above)
Comment (blackswanny) [reply src/slider/__tests__/slider.test.js]: ok

File: src/slider/__tests__/slider.test.js (hunk 5 above)
Comment (blackswanny) [reply src/slider/__tests__/slider.test.js]: the test name explains it `should not move thumb to X pixels if mouse moved less then step`
so mouse move is 20 and it is less than step=50. I added extra text there

File: src/slider/__tests__/stateful-slider.test.js (hunk 6 above)
Comment (blackswanny) [reply src/slider/__tests__/stateful-slider.test.js]: we need to mock it right before we require it in test, so we are able to track mocks for each test. If we bubble require to `imports` jest will start mocking it for the whole file once and forever, so we can't know which test receives which mock call

**Comment by blackswanny** [comment]
> Would be nice to add the check for document.removeEventListener calls. Other than that looks good.


added

**Pull request merged**